benches/pcs.rs

```diff
@@ -0,0 +1,181 @@
+use arecibo::provider::{
+  ipa_pc::EvaluationEngine as IPAEvaluationEngine, mlkzg::EvaluationEngine as MLEvaluationEngine,
+  non_hiding_zeromorph::ZMPCS, Bn256EngineKZG, Bn256EngineZM, GrumpkinEngine,
+};
+use arecibo::spartan::polys::multilinear::MultilinearPolynomial;
+use arecibo::traits::{
+  commitment::CommitmentEngineTrait, evaluation::EvaluationEngineTrait, Engine,
+  TranscriptEngineTrait,
+};
+use criterion::{criterion_group, criterion_main, Bencher, BenchmarkId, Criterion, SamplingMode};
+use halo2curves::bn256::Bn256;
+use rand::rngs::StdRng;
+use rand_core::{CryptoRng, RngCore, SeedableRng};
+use std::time::Duration;
+
+// To run these benchmarks, first download `criterion` with `cargo install cargo-criterion`.
+// Then `cargo criterion --bench pcs`.
+// For flamegraphs, run `cargo criterion --bench pcs --features flamegraph -- --profile-time <secs>`.
+// The results are located in `target/criterion/profile/<name-of-benchmark>`.
+cfg_if::cfg_if! {
+  if #[cfg(feature = "flamegraph")] {
+    criterion_group! {
+          name = pcs;
+          config = Criterion::default().warm_up_time(Duration::from_millis(3000)).with_profiler(pprof::criterion::PProfProfiler::new(100, pprof::criterion::Output::Flamegraph(None)));
+          targets = bench_pcs
+    }
+  } else {
+    criterion_group! {
+          name = pcs;
+          config = Criterion::default().warm_up_time(Duration::from_millis(3000));
+          targets = bench_pcs
+    }
+  }
+}
+
+criterion_main!(pcs);
+
+const TEST_ELL: [usize; 11] = [10, 11, 12, 23, 14, 15, 16, 17, 18, 19, 20];
+
+struct BenchAssests<E: Engine, EE: EvaluationEngineTrait<E>> {
+  poly: MultilinearPolynomial<<E as Engine>::Scalar>,
+  point: Vec<<E as Engine>::Scalar>,
+  eval: <E as Engine>::Scalar,
+  ck: <<E as Engine>::CE as CommitmentEngineTrait<E>>::CommitmentKey,
+  commitment: <<E as Engine>::CE as CommitmentEngineTrait<E>>::Commitment,
+  prover_key: <EE as EvaluationEngineTrait<E>>::ProverKey,
+  verifier_key: <EE as EvaluationEngineTrait<E>>::VerifierKey,
+  proof: Option<<EE as EvaluationEngineTrait<E>>::EvaluationArgument>,
+}
+
+// Macro to generate benchmark code for multiple engine and evaluation engine types
+macro_rules! benchmark_engines {
+    ($ell:expr, $rng:expr, $group:expr, $internal_fn:expr, $( ($engine:ty, $eval_engine:ty, $engine_name:expr) ),*) => {
+        $(
+            let mut assets = deterministic_assets::<$engine, $eval_engine, StdRng>($ell, &mut $rng);
+            $group.bench_with_input(BenchmarkId::new($engine_name, $ell), &$ell, |b, _| {
+                $internal_fn(b, &mut assets);
+            });
+        )*
+    };
+}
+
+fn bench_pcs(c: &mut Criterion) {
+  for &ell in TEST_ELL.iter() {
+    let mut rng = rand::rngs::StdRng::seed_from_u64(ell as u64);
+
+    // Proving group
+    {
+      let mut proving_group = c.benchmark_group("PCS-Proving");
+      proving_group
+        .sampling_mode(SamplingMode::Flat)
```

Comment:
As far as I understand, `Flat` is only suitable to use for very long-running benchmarks. Probably `Auto` can be more appropriate for the short-running ones, where `ell` is say 10 / 11 / 12. Does it make sense to split the benchmarks into short / long running categories?

Reply:
Funny enough: in general the `Flat` sampling mode is fine for most cases ... except I looked at the sample distribution for this PR, and the standard deviation for ZM verification at ell = 10 doesn't look great. You're probably right @storojs72, good catch!

Though I would note one very simple way of addressing this is noting the bench runtimes, which are reasonable (<20s per ell setting, see the bench log on my M2 at https://gist.github.com/huitseeker/015f07534f410e94d615684ecd86876f) re-run them *without* setting the sampling mode (defaulting to `Auto`), and see if the runtimes are still reasonable.

<img width="715" alt="Screenshot 2024-01-04 at 10 03 38 AM" src="https://github.com/lurk-lab/arecibo/assets/4142/c1f7ef32-3152-41aa-b899-a51e022a7295">
<img width="715" alt="Screenshot 2024-01-04 at 10 04 06 AM" src="https://github.com/lurk-lab/arecibo/assets/4142/c48f6e7d-8bea-485a-b3a1-6856a868f97c">
<img width="657" alt="Screenshot 2024-01-04 at 10 04 22 AM" src="https://github.com/lurk-lab/arecibo/assets/4142/b88fb696-3509-4d06-86c0-d8ade1f0dd33">
<img width="709" alt="Screenshot 2024-01-04 at 10 04 30 AM" src="https://github.com/lurk-lab/arecibo/assets/4142/eeb1d10a-a142-4c13-8754-deed5b8b2fd0">

Reply:
As we have removed all odds sizes we are only left with 10 and 12 from your example. I'm not sure that there is a need to split the code for 2 cases. 

I would either make it `Auto` for all (might not be great for higher-end number of variables) or `Flat` (losing a bit of iterations on smaller number of variables but we should still get enough to have convincing results).

Reply:
I'm explicitly suggesting Auto for all, and then revisiting the question if the overall bench runtime is larger than say 15 min.

Reply:
I don't know why I couldn't see your comment when I put mine up. I'll update to `Auto`